cloud/pkg/operationcontroller/controller/downstream.go

```diff
@@ -0,0 +1,357 @@
+		go dc.handleNodeUpgradeTimeout(node, upgrade.Name, upgrade.Spec.Version, newHistoryID, upgrade.Spec.TimeoutSeconds)
+
+		// mark Upgrade state upgrading
+		err = patchNodeUpgradeStatus(dc.crdClient, upgrade, node, v1alpha1.Upgrading, v1alpha1.History{
+			HistoryID:   newHistoryID,
+			UpgradeTime: time.Now().String(),
+		})
+		if err != nil {
+			// not return, continue to mark node unschedulable
+			klog.Errorf("Failed to mark Upgrade upgrading status: %v", err)
+		}
+
+		// mark edge node unschedulable
+		// the effect is like running cmd: kubectl drain <node-to-drain> --ignore-daemonsets
+		unscheduleNode := v1.Node{}
+		unscheduleNode.Spec.Unschedulable = true
+		// add a upgrade label
+		unscheduleNode.Labels = map[string]string{UpgradeStatusKey: UpgradeStatusValue}
+		// add a annotation to mark which Upgrade CR trigger the node upgrade
+		unscheduleNode.Annotations = map[string]string{fmt.Sprintf(UpgradeIDKey, time.Now().UnixNano()): upgrade.Name}
+		byteNode, err := json.Marshal(unscheduleNode)
+		if err != nil {
+			klog.Warningf("marshal data failed: %v", err)
+			continue
+		}
+
+		_, err = dc.kubeClient.CoreV1().Nodes().Patch(context.Background(), node, apimachineryType.StrategicMergePatchType, byteNode, metav1.PatchOptions{})
+		if err != nil {
+			klog.Errorf("failed to drain node %s: %v", node, err)
+			continue
+		}
+	}
+}
+
+func needUpgrade(node *v1.Node, upgradeVersion string) bool {
+	if filterVersion(node.Status.NodeInfo.KubeletVersion, upgradeVersion) {
+		klog.Warningf("Node(%s) version(%s) already on the expected version %s.", node.Name, node.Status.NodeInfo.KubeletVersion, upgradeVersion)
+		return false
+	}
+
+	// we only care about edge nodes, so just remove not edge nodes
+	if !isEdgeNode(node) {
+		klog.Warningf("Node(%s) is not edge node", node.Name)
+		return false
+	}
+
+	// if node is in Upgrading state, don't need upgrade
+	if _, ok := node.Labels[UpgradeStatusKey]; ok {
+		klog.Warningf("Node(%s) is in upgrade state", node.Name)
+		return false
+	}
+
+	// if node is in NotReady state, don't need upgrade
+	for _, condition := range node.Status.Conditions {
+		if condition.Type == v1.NodeReady && condition.Status != v1.ConditionTrue {
+			klog.Warningf("Node(%s) is in NotReady state", node.Name)
+			return false
+		}
+	}
+
+	return true
+}
+
+// handleNodeUpgradeTimeout is used to handle the situation that cloud don't receive upgrade result from edge node
+// within the timeout period
+func (dc *DownstreamController) handleNodeUpgradeTimeout(node string, upgradeID string, upgradeVersion string, historyID int, timeoutSeconds *uint32) {
+	// by default, if we don't receive upgrade response in 300s, we think it's timeout
+	// if we have specified the timeout in Upgrade, we'll use it as the timeout time
+	var timeout uint32 = 300
+	if timeoutSeconds != nil {
+		timeout = *timeoutSeconds
+	}
+	timer := time.NewTimer(time.Duration(timeout) * time.Second)
+	<-timer.C
```

Comment:
Recommend function `wait.PollImmediate(interval, timeout time.Duration, condition ConditionFunc):` PollImmediate tries a condition func until it returns true, an error, or the timeout is reached.